### PR TITLE
Portindex2json: Make the license field consistent with "port info" command

### LINKED ISSUE #44: Format of the License field
Not a super important issue but the license field can be improved sometime in the future to have a consistent format and without braces on the ends.
1. https://ports.macports.org/port/automake
2. https://ports.macports.org/port/aqbanking5

### PR DESCRIPTION
**Pull request by arjunsalyan**
Closes: https://github.com/macports/macports-webapp/issues/44

### CODE REVIEW

**Comment by ryandesign** [comment]
I don't think I agree. Just as the data is in machine-readable format in the PortIndex and is formatted to be human readable in `port info` by MacPorts base, so too it is appropriate for the data to remain in machine-readable format in PortIndex.json and be formatted by the web app.

**Comment by arjunsalyan** [comment]
A TCL list in portindex.json is only as good as a string. We can either make it a list of strings in JSON also or leave it as it is and do everything while rendering in the webapp.

Changing things in portindex.json will anyway require some work on the database also. So, we can close this and move ahead with https://github.com/macports/macports-webapp/pull/235 which formats the license field when rendering the pages (unless there are any other suggestions).

**Comment by ryandesign** [comment]
> A TCL list in portindex.json is only as good as a string.

PortIndex is a Tcl file, so of course it contains Tcl data structures. PortIndex.json is supposed to be a JSON file, so I assumed it would have JSON representations of the data. I see now that for the license field it just contains the Tcl string. That seems wrong to me.

Taking the example of the py-setuptools port, its license in Tcl representation is `MIT {PSF ZPL} {Apache-2 BSD}`. In other words, it is a list of three items. The first item is the string "MIT". The second item is a 2-item list containing the strings "PSF" and "ZPL". The third item is a 2-item list containing the strings "Apache-2" and "BSD". So I'd expect PortIndex.json's license fields to contain whatever the JSON representation of such a nested list (array) structure is, and then I'd expect the web app to parse that and display it in human-readable form.

**Comment by jmroot** [comment]
> Just as the data is in machine-readable format in the PortIndex and is formatted to be human readable in `port info` by MacPorts base, so too it is appropriate for the data to remain in machine-readable format in PortIndex.json and be formatted by the web app.

I tend to agree, that would be a better approach. Among other things, it avoids limiting what else can be done with the JSON.

**Comment by arjunsalyan** [comment]
> Taking the example of the py-setuptools port, its license in Tcl representation is `MIT {PSF ZPL} {Apache-2 BSD}`. In other words, it is a list of three items. The first item is the string "MIT". The second item is a 2-item list containing the strings "PSF" and "ZPL". The third item is a 2-item list containing the strings "Apache-2" and "BSD". So I'd expect PortIndex.json's license fields to contain whatever the JSON representation of such a nested list (array) structure is, and then I'd expect the web app to parse that and display it in human-readable form.

Thank you, I have done the changes and new license field now looks like this in JSON:

```
"license"          : ["MIT",["PSF","ZPL"],["Apache-2","BSD"]]
```

This should be the correct representation in JSON.

I am converting this PR to draft as it should not be merged before the needed changes get deployed on the webapp. I have used recursion to minimize the lines of code, not sure if is  a good idea- that code block could use some feedback.

File: portindex2json/portindex2json.tcl [hunk 1]
```diff
@@ -142,7 +147,7 @@ while {[gets $fd line] >= 0} {
     array set portinfo $line
     array unset json_portinfo
     foreach key [array names portinfo] {
-        if {$key eq "categories" || $key eq "variants" || [string match "depends_*" $key]} {
+        if {$key eq "categories" || $key eq "variants" || $key eq "license" || [string match "depends_*" $key]} {
```
Comment (ryandesign) [inline portindex2json/portindex2json.tcl]: By the way you can use the `in` operator here, as in:
```
if {$key in [list categories variants license] || [string match "depends_*" $key]} {
```

**Review by ryandesign — APPROVED** [review]
Thanks! It looks good to me. I don't know whether the recursion could be a problem. Not sure how Tcl handles recursion. But assuming it handles it at all, which it seems to, the license field won't be that complex that it would cause very much recursion.

**Comment by arjunsalyan** [comment]
This can now be merged unless there is any feedback.

**Review by jmroot — APPROVED** [review]
LGTM

**Comment by mojca** [comment]
So just to make it clear, here's how we would want the Tcl transformed (`Tcl` => `json` => `human-readable form`)?

* `MIT` => `["MIT"]` => `MIT`
* `MIT BSD` => `["MIT", "BSD"]` => `MIT and BSD`
* `{MIT BSD}` => `[["MIT", "BSD"]]` => `MIT or BSD`
* `MIT {PSF ZPL} {Apache-2 BSD}` => `["MIT", ["PSF", "ZPL"], ["Apache-2", "BSD"]]` => `MIT and (PSF or ZPL) and (Apache-2 or BSD)`

**Comment by ryandesign** [comment]
Yes (except "(Apache-2 *or* BSD)").

**Comment by mojca** [comment]
> Yes (except "(Apache-2 _or_ BSD)").

Oh, sorry, yes (question fixed).
We probably need an unit test for this inside the webapp.

**Pull request merged**